### PR TITLE
[flutter_tool] Reland: support --fast-start for Android applications (as an opt-in)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

**TLDR**: Reduces the time it takes to get from `flutter run` to an interactive debugging session. Locally this reduces the startup time for a dirty application from ~40 seconds in the flutter gallery to about ~12 seconds.

The default behavior of flutter run is to build the platform binary (APK, IPA) as necessary and update it on the device before starting a debug session. This artifact is almost always invalidated whenever the user restarts a debugging session, because it depends on all Dart sources, assets, and native dependencies. Depending on the speed of the user's workstation, this can take anywhere from several seconds to several minutes to rebuild.

Both Dart code and Flutter assets can be loaded at runtime from the devfs instead of from the APK - this is how hot restart works. If the Dart source code and assets are removed as inputs from the APK, then invalidations will only occur on Flutter version updates or plugin changes, which will be significantly less frequent than source code changes.

Instead of building the user's application in the initial APK, we build the empty "splash" application which looks like so:

After syncing files, the tool performs a silent hot restart to begin the debugging session.

~~Known issues: IDEs are not supported because we cannot hot restart into a paused state https://github.com/flutter/flutter/issues/45424~~

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Currently testing if build-filters are the cause of the skipped tests

**Comment by jonahwilliams** [comment]
Testing that test fails...

**Comment by jonahwilliams** [comment]
Removed CI/build changes

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -211,6 +211,15 @@ class HotRunner extends ResidentRunner {
       }
     }
 
+    if (debuggingOptions.fastStart) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/run_hot.dart]: Consider a comment here, like "In fastStart mode, apps are initialized from a placeholder splashscreen app. We must do a restart here to load the program and assets for the real app".

**Review by zanderso — APPROVED** [review]
lgtm w/ suggestion for a comment.

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/run_hot.dart]: Done

**Comment by jonahwilliams** [comment]
@feinstein if you'd like to try this out, its available on the latest master. Remember, to see the improvements you have to build the APK with the flag at least once

**Comment by feinstein** [comment]
Ok, but can't promise I will as I tend to avoid changing my flutter framework to avoid headaches.

**Comment by Hixie** [comment]
How did this address the issues I raised with it last time?

**Pull request merged**